numba/core/untyped_passes.py

```diff
@@ -1680,6 +1681,103 @@ def run_pass(self, state):
+            return False
+        call_fn = get_definition(func_ir, call_expr.func)
+
+        if call_fn.value not in (iter, zip):
+            return False
+
+        for arg in call_expr.args:
+            # only one arg?
+            if self.is_arg_constsized(state, arg):
+                # store the call expr to be removed later
+                self.dead_vars.append(func_ir.get_assignee(call_expr))
+                return True
+        return False
+
+    def find_dead_loops(self, state):
+        dead_loops = list()
+        cfg = compute_cfg_from_blocks(state.func_ir.blocks)
+        for loop in cfg.loops().values():
+            header = state.func_ir.blocks[loop.header]
+            for expr in header.find_exprs('iternext'):
+                expr = guard(get_definition, state.func_ir, expr.value.name)
+                if expr:
+                    if isinstance(expr, ir.Expr) and expr.op == 'getiter':
+                        # expr := getiter(value=Var)
+                        if self.is_arg_constsized(state, expr.value) or \
+                           self.match_iter(state, expr):
+                            dead_loops.append(loop)
+        return dead_loops
+
+    def remove_dead_loops(self, state, dead_loops):
+        for loop in dead_loops:
+            # change branch into a direct jump to exit
+            header = state.func_ir.blocks[loop.header]
+            term = header.terminator
+            header.insert_before_terminator(ir.Jump(term.falsebr, term.loc))
+            header.remove(term)
+
+            # iterate over header and remove "pair_first", since
+            # dead_code_elimination won't remove it
+            for inst in header.find_insts(ir.Assign):
+                expr = inst.value
+                if isinstance(expr, ir.Expr) and expr.op == 'pair_first':
+                    header.remove(inst)
+
+        dead_branch_prune(state.func_ir, state.args)
+
+        for var in self.dead_vars:
+            if var is None:
+                continue
+            for block in state.func_ir.blocks.values():
+                assign = block.find_variable_assignment(var.name)
+                if assign:
+                    block.remove(assign)
+
+    def run_pass(self, state):
+        # run as subpipeline
+        from numba.core.compiler_machinery import PassManager
+        from numba.core.typed_passes import PartialTypeInference
+        pm = PassManager("dead_loop_elimination_subpipeline")
+        if not state.flags.no_rewrites:
+            pm.add_pass(RewriteSemanticConstants, "Semantic const")
+            pm.add_pass(DeadBranchPrune, "Dead branch prune")
+            pm.add_pass(GenericRewrites, "Generic Rewrites")
+        pm.add_pass(PartialTypeInference, "performs partial type inference")
+        pm.finalize()
+        pm.run(state)
+
+        dead_loops = self.find_dead_loops(state)
+        if dead_loops:
+            self.remove_dead_loops(state, dead_loops)
```

Comment:
I think this ends up with an invalid intermediate IR as there's a `phi` referring to the target of the `pair_first`, it just so happens that DCE removes it, but it is invalid. Example:
```python
from numba import njit
import numpy as np

@njit
def foo():
    for x in ():
        pass
foo()
```
has this state at this point:
```
label 0:
    $const2.0 = build_tuple(items=[])        ['$const2.0']
    $4get_iter.1 = getiter(value=$const2.0)  ['$4get_iter.1', '$const2.0']
    $phi6.0 = $4get_iter.1                   ['$4get_iter.1', '$phi6.0']
    jump 6                                   []
label 6:
    $6for_iter.1 = iternext(value=$phi6.0)   ['$6for_iter.1', '$phi6.0']
    $6for_iter.3 = pair_second(value=$6for_iter.1) ['$6for_iter.1', '$6for_iter.3']
    $phi8.1 = $6for_iter.2                   ['$6for_iter.2', '$phi8.1'] # THIS REFERS TO SOMETHING THAT DOES NOT EXIST!
    jump 12                                  []
label 12:
    $const12.0 = const(NoneType, None)       ['$const12.0']
    $14return_value.1 = cast(value=$const12.0) ['$14return_value.1', '$const12.0']
    return $14return_value.1                 ['$14return_value.1']
```

Reply:
But DCE will make the IR valid again, no? This is the IR I get after dead code elimination

```
label 0:
    $const12.0 = const(NoneType, None)       ['$const12.0']
    $14return_value.1 = cast(value=$const12.0) ['$14return_value.1', '$const12.0']
    return $14return_value.1                 ['$14return_value.1']
```

Reply:
It does happen to "fix it", but I wonder if this is always the case?